src/utils/formatting.tsx

```diff
@@ -0,0 +1,59 @@
+import React, { ReactElement, ReactNode } from "react";
+import styled from "styled-components";
+
+export const RenderedStringContainer = styled.div`
+  display: flex;
+  flex-direction: column;
+  gap: 2px;
+
+  & > p,
+  & > ul {
+    margin: 0;
+  }
+`;
+
+/**
+ * Renders a text string array as JSX react elements, handling formatting for lists and paragraph text.
+ * @param items List of string text items to render.
+ * @param containerStyle optional CSS properties object that will be applied to the container div.
+ * @returns A list of one or more text elements, wrapped in a styling div.
+ * String items will be returned as one of the following:
+ * - If it starts with "- ", the item will be rendered as a list item (`li`) element. Groups of `li` elements
+ * will be wrapped in a unordered list (`ul`) element.
+ * - Otherwise, the item is wrapped in a `p` element.
+ */
+export function renderStringArrayAsJsx(items: string[] | string | undefined): ReactElement {
+  if (!items || items.length === 0) {
+    return <></>;
+  }
+  if (typeof items === "string") {
+    return (
+      <RenderedStringContainer>
+        <p>{items}</p>;
+      </RenderedStringContainer>
+    );
+  }
+
+  const elements: ReactNode[] = [];
+  let currListElements: ReactNode[] = [];
+
+  for (let i = 0; i < items.length; i++) {
+    const item = items[i].trim();
+    if (item.startsWith("- ")) {
+      currListElements.push(<li key={currListElements.length}>{item.substring(2)}</li>);
+      continue;
+    } else {
+      if (currListElements.length > 0) {
+        elements.push(<ul key={i - 1}>{currListElements}</ul>);
+        currListElements = [];
+      }
+      elements.push(<p key={i}>{item}</p>);
+    }
+  }
```

Comment:
This is a neat utility. It's only called in a couple places and the behavior is pretty bespoke, so given the amount of test code that's necessary I'm curious the rationale for adding it, as opposed to wrapping `map` calls in the new `RenderedStringContainer`?

Non-blocking, just curious.

Reply:
The old map calls didn't render list items as `li` or `ul` elements, just as `p` elements, so we would have lost some of implicit HTML styling/accessibility attributes we get from lists. I had gotten some feedback from UX about having the text look more like bullet points, so I figured it would be better to explicitly transform them into HTML unordered lists.

If this ever needs to handle more cases it'll probably be better to use a Markdown to HTML renderer.

Test code was optional but figured I might as well throw it in!

Reply:
Thanks for context! Always keeping an eye on your custom tooling, for use in other repos, especially in V2 era.